art/attacks/evasion/simba.py

```diff
@@ -0,0 +1,319 @@
+                        current_label = np.argmax(right_preds, axis=1)[0]
+            
+            if self.targeted:
+                if desired_label == current_label:
+                    term_flag = 1
+            else:
+                if desired_label != current_label:
+                    term_flag = 1
+
+            nb_iter = nb_iter + 1
+
+        if nb_iter < self.max_iter:
+            logger.info('SimBA (%s) %s attack succeed', self.attack, ['non-targeted', 'targeted'][int(self.targeted)])
+        else:
+            logger.info('SimBA (%s) %s attack failed', self.attack, ['non-targeted', 'targeted'][int(self.targeted)])
+
+        return x
+
+    def set_params(self, **kwargs):
+        """
+        Take in a dictionary of parameters and applies attack-specific checks before saving them as attributes.
+
+        :param attack: attack type: pixel (px) or DCT (dct) attacks
+        :type attack: `str`
+        :param max_iter: The maximum number of iterations.
+        :type max_iter: `int`
+        :param epsilon: Overshoot parameter.
+        :type epsilon: `float`
+        :param order: order of pixel attacks
+        :type order: `str`
+        :param freq_dim: dimensionality of 2D frequency space (DCT).
+        :type freq_dim: `int`
+        :param stride: stride for block order (DCT).
+        :type stride: `int`
+        :param targeted: targeted attacks
+        :type targeted: `bool`
+        :param batch_size: Batch size (but, batch process unavailable in this implementation)
+        :type batch_size: `int`
+        """
+        # Save attack-specific parameters
+        super(SimBA, self).set_params(**kwargs)
+
+        if not isinstance(self.max_iter, (int, np.int)) or self.max_iter <= 0:
+            raise ValueError("The number of iterations must be a positive integer.")
+
+        if self.epsilon < 0:
+            raise ValueError("The overshoot parameter must not be negative.")
+
+        if self.batch_size != 1:
+            raise ValueError('The batch size `batch_size` has to be 1 in this implementation.')
+        
+        if not isinstance(self.stride, (int, np.int)) or self.stride <= 0:
+            raise ValueError("The `stride` value must be a positive integer.")
+        
+        if not isinstance(self.freq_dim, (int, np.int)) or self.freq_dim <= 0:
+            raise ValueError("The `freq_dim` value must be a positive integer.")
+        
+        if self.order != 'random' and self.order != 'diag':
+            raise ValueError('The order of pixel attacks has to be `random` or `diag`.')
+        
+        if self.attack != 'px' and self.attack != 'dct':
+            raise ValueError('The attack type has to be `px` or `dct`.')
+        
+        if not isinstance(self.targeted, (int)) or (self.targeted != 0 and self.targeted != 1):
+            raise ValueError('`targeted` has to be a logical value.')
+
+        return True
+
+    def _block_order(self, img_size, channels, initial_size=2, stride=1):
```

Comment:
Please add a docstring and typing information to this method.

Reply:
added a docstring and typing information to this method.